Bio/SeqIO/QualityIO.py

```diff
@@ -1082,23 +1086,39 @@ def __next__(self) -> SeqRecord:
                 raise ValueError("Unexpected end of file")
             self.line = None
 
-        if seq_len != len(quality_string):
-            raise ValueError(
-                "Lengths of sequence and quality values differs for %s (%i and %i)."
-                % (title_line, seq_len, len(quality_string))
-            )
         descr = title_line
         id = descr.split()[0]
         name = id
-        q_mapping = self.q_mapping
-        try:
-            qualities = [q_mapping[letter2] for letter2 in quality_string]
-        except KeyError:
-            raise ValueError("Invalid character in quality string") from None
+
+        if not quality_string.isascii():
+            # Look for invalid non-ascii characters
+            index = _find_index_where(quality_string, lambda c: not c.isascii())
+            assert index >= 0, "Non-ascii char in qualities not found. Biopython bug?"
+
+            details = "is not an ASCII character"
+            raise ValueError(_invalid_char_msg(index, quality_string, details))
+
+        if len(quality_string) != seq_len:
+            # This should happen after ascii check, because non-ascii characters will often trigger this
+            raise ValueError(
+                f"Lengths of sequence and quality values differs for {title_line} ({seq_len} and {len(quality_string)})."
+            )
+
+        byte_scores = quality_string.encode().translate(self.q_mapping)
+
+        if INVALID_CHAR in byte_scores:
+            # Look for invalid ascii characters
+            invalid_index = byte_scores.find(INVALID_CHAR_CODE)
+
+            details = "not in correct range (are you sure you're using the right QualityIO parser?)"
+            raise ValueError(_invalid_char_msg(invalid_index, quality_string, details))
+
+        # Pass through (standard library) array to handle negative scores from old quality formats
+        qualities = array.array("b", byte_scores).tolist()
 
         # # Avoid length/type checking
         record = SeqRecord._from_validated(
-            Seq(seq_string),
+            Seq(seq_string.encode()),
```

Comment:
The `.encode()` is not needed here; the `Seq` initializer will take care of it.

Reply:
True, but doing it here saves a few if-statements, so should be faster - and we do care about that for FASTQ. It might even be measurable?

Reply:
Yeah, it's a fair bit faster. Encode + Seq init is about 2.3s per 10 million calls, compared to passing in a string which is about 6.6s per million calls. I think this is because with Bytes it only has to do 1 or 2 isinstance checks, while string has to do 5. 

If bytearrays are used less frequently than strings, it might be worth just moving up the string check to be third instead of fourth, though pre-encoding is probably going to remain faster

Reply:
It's also possible that we might want to make custom constructors like from_bytes or from_str if Seq construction is ever a bottleneck. I think for this case we're only talking maybe like 5% of the runtime so probably not worth it

Reply:
OK. Maybe just leave a comment then on this line to explain to the next developer looking at this why we convert to bytes here.

Reply:
Yeah, definitely. I could have sworn I did at some point, but it must have gotten lost in one of the merges